domains/client/domain-operator/src/snap_sync.rs

```diff
@@ -0,0 +1,481 @@
+    );
+
+    // Clear the block gap that arises from first block import with a much higher number than
+    // previously (resulting in a gap)
+    // TODO: This is a hack and better solution is needed: https://github.com/paritytech/polkadot-sdk/issues/4407
+    sync_params.domain_client.clear_block_gap()?;
+
+    if let Some(offchain_storage) = sync_params
+        .consensus_chain_sync_params
+        .backend
+        .offchain_storage()
+    {
+        // let target_block = sync_params
+        //     .consensus_chain_sync_params
+        //     .segment_headers_store
+        //     .last_segment_header()
+        //     .map(|header| header.last_archived_block().number);
+
+        let target_block = sync_params
+            .consensus_chain_sync_params
+            .segment_headers_store
+            .max_segment_index()
+            // Skip last `FINALIZATION_DEPTH_IN_SEGMENTS` archived segments
+            .and_then(|max_segment_index| {
+                max_segment_index.checked_sub(FINALIZATION_DEPTH_IN_SEGMENTS)
+            })
+            .and_then(|segment_index| {
+                sync_params
+                    .consensus_chain_sync_params
+                    .segment_headers_store
+                    .get_segment_header(segment_index)
+            })
+            .map(|segment_header| segment_header.last_archived_block().number);
+
+        mmr_sync(
+            sync_params.consensus_chain_sync_params.fork_id,
+            Arc::clone(&sync_params.consensus_client),
+            sync_params.consensus_chain_sync_params.network_service,
+            Arc::clone(&sync_params.consensus_chain_sync_params.sync_service),
+            offchain_storage,
+            target_block,
+        )
+        .await?;
+    }
+
+    sync_params
+        .consensus_chain_sync_params
+        .snap_sync_orchestrator
+        .mark_domain_snap_sync_finished();
+
+    debug!(info = ?sync_params.domain_client.info(), "Client info after successful domain snap sync.");
+
+    Ok(())
+}
+
+/// Download and return state for specified block
+async fn download_state<Block, Client, NR>(
+    header: &Block::Header,
+    client: &Arc<Client>,
+    fork_id: Option<String>,
+    network_request: &NR,
+    sync_service: &SyncingService<Block>,
+) -> Result<ImportedState<Block>, sp_blockchain::Error>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block> + ProofProvider<Block> + Send + Sync + 'static,
+    NR: NetworkRequest,
+{
+    let block_number = *header.number();
+
+    const STATE_SYNC_RETRIES: u32 = 5;
+    const LOOP_PAUSE: Duration = Duration::from_secs(20);
+    const MAX_GET_PEERS_ATTEMPT_NUMBER: usize = 30;
+
+    for attempt in 1..=STATE_SYNC_RETRIES {
+        debug!(%block_number, %attempt, "Starting state sync...");
+
+        debug!(%block_number, "Gathering peers for state sync.");
+        let mut tried_peers = HashSet::<PeerId>::new();
+
+        let mut current_get_peers_attempt = 0usize;
+        let current_peer_id = loop {
```

Comment:
It seems `get_last_confirmed_block` has this same piece of logic it would be great to reuse it.